src/pudl/output/ferc1.py

```diff
@@ -1337,3 +1339,93 @@ def remove_intra_table_calculated_values(exploded: pd.DataFrame) -> pd.DataFrame
         )
     ]
     return exploded
+
+
+################################################################################
+# XBRL Calculation Tree
+################################################################################
+Ferc1XbrlCalculationComponent = ForwardRef("Ferc1XbrlCalculationComponent")
+Ferc1XbrlCalculation = ForwardRef("Ferc1XbrlCalculation")
+
+
+class Ferc1XbrlCalculationComponent(BaseModel):
+    """A class representing the components of a FERC 1 XBRL Metadata calculation."""
+
+    name: str
+    weight: float
+    source_tables: list[str]
+    calculation: Ferc1XbrlCalculation | None = None
+
+    @validator("source_tables")
+    def has_unique_source_table(cls, value):
+        """Ensure that each calculation component has a unique source table."""
+        if len(value) == 1:
+            return value
+        else:
+            raise ValueError(
+                "Exploding tables requires calculations to have unique source tables."
+            )
+
+
+class Ferc1XbrlCalculation(BaseModel):
+    """A class representing a calculation from the FERC 1 XBRL Metadata."""
+
+    components: list[Ferc1XbrlCalculationComponent] = []
+    source_table: str
+    xbrl_factoid: str
+    xbrl_factoid_original: str
+
+    @property
+    def referenced_tables(self: Self) -> set[str]:
+        """Identify the set of all referenced tables in a calculation."""
+        return {c for x in self.calculations for c in x.source_tables}
+
+    def is_inter_table(self: Self, source_table: str | None = None) -> bool:
+        """Determine whether a calculation involves more than one table."""
+        all_tables = self.referenced_tables
+        if source_table is not None:
+            all_tables = all_tables.union({source_table})
+        return True if len(all_tables) > 1 else False
+
+
+Ferc1XbrlCalculationComponent.update_forward_refs()
+Ferc1XbrlCalculation.update_forward_refs()
+
+
+def resolve_calculation_components(
```

Comment:
I'm having a little bit of trouble figuring out what the expectations of this function are, mostly because I don't know much about what the exploded metadata should look like, or what we expect these calculation trees to look like!

I started writing some tests to try to nail down some of that behavior, they look like this so far - it might make sense for someone who knows more about the expected behavior to actually flesh these out.

From my perspective having these tests will give a lot of confidence that this calculation tree will have the shape we expect, but if you feel like the tests would be redundant I'm happy to be convinced to have less code :)

```python
import json
import pandas as pd

from pudl.output.ferc1 import resolve_calculation_components


def test_resolve_calculation_components_empty():
    # maybe there's a better way to get an exploded test metadata than just hand-coding
    # this structure...
    test_table = "test_table"
    test_factoid = "test_factoid"
    calculations = json.dumps([])
    exploded_meta = pd.DataFrame({
        "table_name": [test_table],
        "xbrl_factoid": [test_factoid],
        "xbrl_factoid_original": [f"{test_factoid}_original"],
        "calculations": [calculations],
    })
    exploded_meta = exploded_meta.set_index(["table_name", "xbrl_factoid"])

    observed = resolve_calculation_components(exploded_meta, test_table, test_factoid)

    assert observed.source_table == test_table
    assert observed.xbrl_factoid == test_factoid
    assert observed.xbrl_factoid_original == f"{test_factoid}_original"
    assert observed.components == []


def test_resolve_calculation_components_single():
    pass


def test_resolve_calculation_components_one_level():
    pass


def test_resolve_calculation_components_multi_level():
    pass


def test_resolve_calculation_components_multi_level_non_unique_source_tables():
    pass
```

Reply:
Yeah... it definitely needs some tests.

The high level idea is that we want to expand all of the individual calculation components which are actually references to other values and expose the dependencies between them so that we can see how they nest within each other, and apply metadata from one node to all of its children. E.g. if the `utility_plant_net` fact has been tagged as being "in rate base" then all of its child components should also get tagged with "in rate base" -- then this information can be merged onto the exploded output table, without needing to manually compile and update a huge table that applies every possible tag to every possible fact within one of these big explosions (that involve hundreds of facts).